File: src/gnome.c
/* gnome.c -- minimal GNOME compliance
   $Id$

   Copyright (C) 1999 John Harper <john@dcs.warwick.ac.uk>

   This file is part of sawmill.

   sawmill is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   sawmill is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with sawmill; see the file COPYING.   If not, write to
   the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA. */

/* todo:

   * this assumes knowledge about the workspace implementation; not
     a good idea.. */

#include "sawmill.h"
#include <X11/Xutil.h>
#include <X11/Xutil.h>
#include <alloca.h>

DEFSYM(gnome, "gnome");
DEFSYM(enter_workspace_hook, "enter-workspace-hook");
DEFSYM(add_workspace_hook, "add-workspace-hook");
DEFSYM(delete_workspace_hook, "delete-workspace-hook");
DEFSYM(add_to_workspace_hook, "add-to-workspace-hook");
DEFSYM(remove_from_workspace_hook, "remove-from-workspace-hook");
DEFSYM(ws_workspaces, "ws-workspaces");
DEFSYM(ws_current_workspace, "ws-current-workspace");
DEFSYM(select_workspace, "select-workspace");
DEFSYM(ignored_window_names, "ignored-window-names");
DEFSYM(sticky_window_names, "sticky-window-names");
DEFSYM(ignored, "ignored");

DEFSTRING(gnome_windows, "^(gmc|panel)$");

static Window gnome_id_win;

static Atom xa_win_supporting_wm_check, xa_win_client_list, xa_win_protocols,
    xa_win_workspace, xa_win_workspace_count, xa_win_desktop_button_proxy;


/* client list support */

DEFUN("gnome-set-client-list", Fgnome_set_client_list,
      Sgnome_set_client_list, (void), rep_Subr0)
{
    Lisp_Window *w;
    int count, i;
    Window *clients;

    count = 0;
    for (w = window_list; w != 0; w = w->next)
    {
	repv tem;
	if (w->id == 0 || !w->mapped)
	    continue;
	tem = Fwindow_get (rep_VAL(w), Qignored);
	if (tem && tem != Qnil)
	    continue;
	count++;
    }

    clients = alloca (count * sizeof (Window));
    for (w = window_list, i = 0; w != 0; w = w->next)
    {
	repv tem;
	if (w->id == 0 || !w->mapped)
	    continue;
	tem = Fwindow_get (rep_VAL(w), Qignored);
	if (tem && tem != Qnil)
	    continue;
	clients[i++] = w->id;
    }

    XChangeProperty (dpy, root_window, xa_win_client_list, XA_CARDINAL,
		     32, PropModeReplace, (u_char *)clients, count);
    return Qt;
}


/* workspace support */

DEFUN("gnome-set-workspace", Fgnome_set_workspace,
      Sgnome_set_workspace, (void), rep_Subr0)
{
    repv spaces = Fsymbol_value (Qws_workspaces, Qt);
    repv current = Fsymbol_value (Qws_current_workspace, Qt);
    int total = 0, active = 0;

    while (rep_CONSP(spaces) && !rep_INTERRUPTP)
    {
	repv windows;
	if (rep_CAR(spaces) == current)
	    active = total;
	windows = rep_CDR(rep_CAR(spaces));
	while (rep_CONSP(windows) && !rep_INTERRUPTP)
	{
	    repv win = rep_CAR(windows);
	    if (WINDOWP(win) && VWIN(win)->id != 0)
	    {
		XChangeProperty (dpy, VWIN(win)->id, xa_win_workspace,
				 XA_CARDINAL, 32, PropModeReplace,
				 (u_char *)&total, 1);
	    }
	    windows = rep_CDR(windows);
	}
	total++;
	spaces = rep_CDR(spaces);
	rep_TEST_INT;
    }

    XChangeProperty (dpy, root_window, xa_win_workspace, XA_CARDINAL,
		     32, PropModeReplace, (u_char *)&active, 1);
    XChangeProperty (dpy, root_window, xa_win_workspace_count, XA_CARDINAL,
		     32, PropModeReplace, (u_char *)&total, 1);
    return Qt;
}


/* GNOME client message handling */

DEFUN("gnome-client-message-handler", Fgnome_client_message_handler,
      Sgnome_client_message_handler, (repv win), rep_Subr1)
{
    /* current_x_event is the event that was actually received */
    if (current_x_event == 0)
	return Qnil;
    if (current_x_event->xclient.message_type == xa_win_workspace)
    {
	int desk = current_x_event->xclient.data.l[0];
	rep_call_lisp1 (Qselect_workspace, rep_MAKE_INT(desk));
	return Qt;
    }
    return Qnil;
}


/* event proxying */

static bool
event_proxy (XEvent *ev, long code, long mods)
{
    if (ev->xany.window == root_window
	&& (ev->type == ButtonPress || ev->type == ButtonRelease))
    {
	if (ev->type == ButtonPress)
	    XUngrabPointer (dpy, CurrentTime);
	XSendEvent (dpy, gnome_id_win, False, SubstructureNotifyMask, ev);
	return TRUE;
    }
    else
	return FALSE;
}


/* initialisation */

repv rep_dl_feature;
rep_xsubr *rep_dl_subrs[] = {
    &Sgnome_set_client_list,
    &Sgnome_set_workspace,
    &Sgnome_client_message_handler,
    0
};

static void
add_hook (repv sym, repv fun)
{
    repv val = Fsymbol_value (sym, Qt);
    if (rep_VOIDP(val))
	val = Qnil;
    val = Fcons (fun, val);
    Fset (sym, val);
}

repv
rep_dl_init(repv file_name)
{
    u_int val;
    Atom prot[3];

    rep_INTERN(gnome);
    rep_dl_feature = Qgnome;
    rep_INTERN(enter_workspace_hook);
    rep_INTERN(add_workspace_hook);
    rep_INTERN(delete_workspace_hook);
    rep_INTERN(add_to_workspace_hook);
    rep_INTERN(remove_from_workspace_hook);
    rep_INTERN(ws_workspaces);
    rep_INTERN(ws_current_workspace);
    rep_INTERN(select_workspace);
    rep_INTERN(ignored_window_names);
    rep_INTERN(sticky_window_names);
    rep_INTERN(ignored);

    if (rep_VOIDP(rep_SYM(Qignored_window_names)->value))
	rep_SYM(Qignored_window_names)->value = Qnil;
    if (rep_VOIDP(rep_SYM(Qsticky_window_names)->value))
	rep_SYM(Qsticky_window_names)->value = Qnil;

    rep_SYM(Qignored_window_names)->value
	 = Fcons (rep_VAL(&gnome_windows),
		  rep_SYM(Qignored_window_names)->value);
    rep_SYM(Qsticky_window_names)->value
	 = Fcons (rep_VAL(&gnome_windows),
		  rep_SYM(Qsticky_window_names)->value);

    xa_win_supporting_wm_check = XInternAtom (dpy, "_WIN_SUPPORTING_WM_CHECK", False);
    xa_win_protocols = XInternAtom (dpy, "_WIN_PROTOCOLS", False);
    xa_win_client_list = XInternAtom (dpy, "_WIN_CLIENT_LIST", False);
    xa_win_workspace = XInternAtom (dpy, "_WIN_WORKSPACE", False);
    xa_win_workspace_count = XInternAtom (dpy, "_WIN_WORKSPACE_COUNT", False);
    xa_win_desktop_button_proxy = XInternAtom (dpy, "_WIN_DESKTOP_BUTTON_PROXY", False);

    gnome_id_win = XCreateSimpleWindow (dpy, root_window,
					-200, -200, 5, 5, 0, 0, 0);
    val = gnome_id_win;
    XChangeProperty (dpy, root_window, xa_win_supporting_wm_check,
		     XA_CARDINAL, 32, PropModeReplace, (u_char *)&val, 1);
    XChangeProperty (dpy, gnome_id_win, xa_win_supporting_wm_check,
		     XA_CARDINAL, 32, PropModeReplace, (u_char *)&val, 1);

    XChangeProperty (dpy, root_window, xa_win_desktop_button_proxy,
		     XA_CARDINAL, 32, PropModeReplace, (u_char *)&val, 1);
    XChangeProperty (dpy, gnome_id_win, xa_win_desktop_button_proxy,
		     XA_CARDINAL, 32, PropModeReplace, (u_char *)&val, 1);

    prot[0] = xa_win_client_list;
    prot[1] = xa_win_workspace;
    prot[2] = xa_win_workspace_count;
    XChangeProperty (dpy, root_window, xa_win_protocols,
		     XA_ATOM, 32, PropModeReplace, (u_char *)prot, 3);

    rep_ADD_SUBR(Sgnome_set_client_list);
    rep_ADD_SUBR(Sgnome_set_workspace);
    rep_ADD_SUBR(Sgnome_client_message_handler);

    add_hook (Qenter_workspace_hook, rep_VAL(&Sgnome_set_workspace));
    add_hook (Qadd_workspace_hook, rep_VAL(&Sgnome_set_workspace));
    add_hook (Qdelete_workspace_hook, rep_VAL(&Sgnome_set_workspace));
    add_hook (Qadd_to_workspace_hook, rep_VAL(&Sgnome_set_workspace));
    add_hook (Qremove_from_workspace_hook, rep_VAL(&Sgnome_set_workspace));
    add_hook (Qadd_window_hook, rep_VAL(&Sgnome_set_client_list));
    add_hook (Qdestroy_notify_hook, rep_VAL(&Sgnome_set_client_list));
    add_hook (Qmap_notify_hook, rep_VAL(&Sgnome_set_client_list));
    add_hook (Qunmap_notify_hook, rep_VAL(&Sgnome_set_client_list));
    add_hook (Qclient_message_hook, rep_VAL(&Sgnome_client_message_handler));
    event_proxy_fun = event_proxy;

    return Qt;
}

void
rep_dl_kill(void)
{
    event_proxy_fun = 0;

    XDestroyWindow (dpy, gnome_id_win);
    XDeleteProperty (dpy, root_window, xa_win_supporting_wm_check);
    XDeleteProperty (dpy, root_window, xa_win_protocols);
    XDeleteProperty (dpy, root_window, xa_win_client_list);
    XDeleteProperty (dpy, root_window, xa_win_workspace);
    XDeleteProperty (dpy, root_window, xa_win_workspace_count);
}

File: lisp/sawfish/wm/state/gnome.jl
;; gnome.jl -- minimal GNOME compliance
;; $Id$

;; Copyright (C) 1999 John Harper <john@dcs.warwick.ac.uk>

;; This file is part of sawmill.

;; sawmill is free software; you can redistribute it and/or modify it
;; under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 2, or (at your option)
;; any later version.

;; sawmill is distributed in the hope that it will be useful, but
;; WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with sawmill; see the file COPYING.  If not, write to
;; the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.

(provide 'gnome)

(defconst WIN_STATE_STICKY 1)

(defvar gnome-window-names "^(gmc|panel)$")

(defvar gnome-window-id nil)

(defvar gnome-supported-protocols [_WIN_CLIENT_LIST _WIN_WORKSPACE
				   _WIN_WORKSPACE_COUNT _WIN_STATE])

(defun gnome-set-client-list ()
  (let
      (clients vec)
    (mapc #'(lambda (w)
	      (when (and (windowp w) (window-mapped-p w)
			 (not (window-get w 'ignored)))
		(setq clients (cons (window-id w) clients))))
	  (managed-windows))
    (setq vec (apply 'vector clients))
    (set-x-property 'root '_WIN_CLIENT_LIST vec 'CARDINAL 32)))

(defun gnome-set-workspace ()
  (let
      ((total 0))
    (mapc #'(lambda (space)
	      (when (eq space ws-current-workspace)
		(set-x-property 'root '_WIN_WORKSPACE
				(vector total) 'CARDINAL 32))
	      (mapc #'(lambda (w)
			(set-x-property w '_WIN_WORKSPACE
					(vector total) 'CARDINAL 32))
		    (cdr space))
	      (setq total (1+ total)))
	  ws-workspaces)
    (set-x-property 'root '_WIN_WORKSPACE_COUNT (vector total) 'CARDINAL 32)))

(defun gnome-set-client-state (w)
  (let
      ((state 0))
    (when (window-get w 'sticky)
      (setq state (logior state WIN_STATE_STICKY)))
    (set-x-property w '_WIN_STATE (vector state) 'CARDINAL 32)))

(defun gnome-honour-client-state (w)
  (let
      ((state (get-x-property w '_WIN_STATE)))
    (when (eq (car state) 'CARDINAL)
      (unless (zerop (logand (aref (nth 2 state) 0) WIN_STATE_STICKY))
	(unless (window-get w 'sticky)
	  (toggle-window-sticky w))))
    w))

(defun gnome-client-message-handler (w type data)
  (cond ((eq type '_WIN_WORKSPACE)
	 (select-workspace (aref data 0))
	 t)
	((eq type '_WIN_STATE)
	 (let
	     ((mask (aref data 0))
	      (values (aref data 1))
	      tem)
	   (unless (zerop (logand mask WIN_STATE_STICKY))
	     (setq tem (window-get w 'sticky))
	     (if (or (and (not tem)
			  (not (zerop (logand values WIN_STATE_STICKY))))
		     (and tem (zerop (logand values WIN_STATE_STICKY))))
		 (toggle-window-sticky w))))
	 t)))

(defun gnome-event-proxyer ()
  (when (eq (current-event-window) 'root)
    (let
	((event (event-name (current-event))))
      ;; XXX should check that there's no binding of the
      ;; XXX corresponding event in the root-window-keymap or
      ;; XXX the global-keymap
      (when (string-match "-(Click1|Off)$" event)
	(proxy-current-event gnome-window-id)
	t))))


;; initialisation

(defun gnome-init ()
  (setq ignored-window-names (cons gnome-window-names ignored-window-names))
  (setq sticky-window-names (cons gnome-window-names sticky-window-names))
  (setq gnome-window-id (create-window 'root -200 -200 5 5))

  (set-x-property 'root '_WIN_SUPPORTING_WM_CHECK
		  (vector gnome-window-id) 'CARDINAL 32)
  (set-x-property gnome-window-id '_WIN_SUPPORTING_WM_CHECK
		  (vector gnome-window-id) 'CARDINAL 32)

  (set-x-property 'root '_WIN_DESKTOP_BUTTON_PROXY
		  (vector gnome-window-id) 'CARDINAL 32)
  (set-x-property gnome-window-id '_WIN_DESKTOP_BUTTON_PROXY
		  (vector gnome-window-id) 'CARDINAL 32)

  (set-x-property 'root '_WIN_PROTOCOLS
		  gnome-supported-protocols 'ATOM 32)

  (delete-x-property 'root '_WIN_AREA)
  (delete-x-property 'root '_WIN_AREA_COUNT)
  (delete-x-property 'root '_WIN_WORKSPACE_NAMES)

  (add-hook 'enter-workspace-hook 'gnome-set-workspace)
  (add-hook 'add-workspace-hook 'gnome-set-workspace)
  (add-hook 'delete-workspace-hook 'gnome-set-workspace)
  (add-hook 'add-to-workspace-hook 'gnome-set-workspace)
  (add-hook 'remove-from-workspace-hook 'gnome-set-workspace)

  (add-hook 'add-window-hook 'gnome-set-client-list)
  (add-hook 'destroy-notify-hook 'gnome-set-client-list)
  (add-hook 'map-notify-hook 'gnome-set-client-list)
  (add-hook 'umap-notify-hook 'gnome-set-client-list)

  (add-hook 'client-message-hook 'gnome-client-message-handler)

  (add-hook 'window-state-change-hook 'gnome-set-client-state)
  (add-hook 'add-window-hook 'gnome-honour-client-state)

  (add-hook 'unbound-key-hook 'gnome-event-proxyer)

  (add-hook 'before-exit-hook 'gnome-exit))

(defun gnome-exit ()
  (destroy-window gnome-window-id)
  (delete-x-property 'root '_WIN_SUPPORTING_WM_CHECK)
  (delete-x-property 'root '_WIN_PROTOCOLS))

(unless gnome-window-id
  (gnome-init))
